File: app/src/main/java/edu/SpaceLearning/SpaceEnglish/_Navfragments/QuizNavFragment.java
package edu.SpaceLearning.SpaceEnglish._Navfragments;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.TextView;

import java.util.ArrayList;

import edu.SpaceLearning.SpaceEnglish.Listeners.InteractionActivityFragmentsListener;
import edu.SpaceLearning.SpaceEnglish.QuizFrags.QuizCategoriesFragment;
import edu.SpaceLearning.SpaceEnglish.R;
import edu.SpaceLearning.SpaceEnglish.UtilsClasses.Constants;
import edu.SpaceLearning.SpaceEnglish.UtilsClasses.Scores;
import edu.SpaceLearning.SpaceEnglish.UtilsClasses.Utils;
import edu.SpaceLearning.SpaceEnglish.databinding.FragmentNavQuizBinding;


public class QuizNavFragment extends Fragment  {

    private FragmentNavQuizBinding binding;
    private InteractionActivityFragmentsListener interactionListener;
    private final ArrayList<Button> buttons = new ArrayList<>();
    private final ArrayList <TextView> textViews = new ArrayList<>();

    public QuizNavFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_nav_quiz, container, false);

    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
         binding = FragmentNavQuizBinding.bind(view);


         addUIToListAndInitIt();
         initAutoTextViewMaxQuestionPerQuiz();

    }



    private void initAutoTextViewMaxQuestionPerQuiz() {
        binding.autoTvmaxQuestionsPerQuiz.setText(String.valueOf(Utils.maxQuestionsPerQuiz));

        if(!binding.autoTvmaxQuestionsPerQuiz.getText().toString().isEmpty()) {
            Utils.maxQuestionsPerQuiz = Integer.parseInt(binding.autoTvmaxQuestionsPerQuiz.getText().toString());
        }

        String[] maxQuestionsList = getResources().getStringArray(R.array.max_questions_per_quiz);
        ArrayAdapter<String> adapterAutoCompleteTv = new ArrayAdapter<>(requireActivity(), android.R.layout.simple_list_item_1,maxQuestionsList);
        binding.autoTvmaxQuestionsPerQuiz.setAdapter(adapterAutoCompleteTv);

        binding.autoTvmaxQuestionsPerQuiz.setOnItemClickListener((parent, view1, position, id) -> {
            String txtSelected = binding.autoTvmaxQuestionsPerQuiz.getText().toString();
            Utils.maxQuestionsPerQuiz = Integer.parseInt(txtSelected);
        });
    }

    private void addUIToListAndInitIt() {
        buttons.add(binding.btnVerbs);buttons.add(binding.btnSentences);buttons.add(binding.btnPhrasals);buttons.add(binding.btnNouns);
        buttons.add(binding.btnAdjs);buttons.add(binding.btnAdvs);buttons.add(binding.btnIdioms);
        textViews.add(binding.tvPhrasalRequiredLabel);textViews.add(binding.tvNounRequiredLabel);textViews.add(binding.tvAdjRequiredLabel);
        textViews.add(binding.tvAdvRequiredLabel);textViews.add(binding.tvIdiomRequiredLabel);

        for(int i = 0 ; i<Constants.permissionCategoryScoreArray.length;i++) {
            enableButton(buttons.get(i + 2),textViews.get(i), Scores.totalScore,Constants.permissionCategoryScoreArray[i]);
        }

        for (int i =0 ; i< buttons.size();i++) {
            int finalI = i;
            buttons.get(i).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    interactionListener.onSetRequiredCategoryFragmentQuiz(QuizCategoriesFragment.getInstance(Constants.categoryNameArray[finalI]));
                }
            });
        }
    }

    private void enableButton(Button button , TextView tvRequiredLabel , int globalMainScore , int permissionScore) {

        if(globalMainScore >= permissionScore) {
            button.setEnabled(true);
            tvRequiredLabel.setVisibility(View.GONE);
        }
        else {
            button.setEnabled(false);
            tvRequiredLabel.setVisibility(View.VISIBLE);
        }
    }


    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if(context instanceof InteractionActivityFragmentsListener){
            interactionListener = (InteractionActivityFragmentsListener) context;
        }

    }

    @Override
    public void onDetach() {
        super.onDetach();
        if (interactionListener != null) {
            interactionListener = null;}

    }


}

File: app/src/main/java/edu/SpaceLearning/SpaceEnglish/_Navfragments/QuizNavFragment.kt
/*
 * File: QuizNavFragment.java
 * Author: [Your Name]
 * Date: [Date]
 * Purpose: Fragment representing quiz navigation in the SpaceEnglish app.
 *          Provides UI for selecting quiz categories, setting quiz parameters,
 *          and interaction handling with RecyclerView.
 */
package edu.SpaceLearning.SpaceEnglish._Navfragments

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import edu.SpaceLearning.SpaceEnglish.Adapters.RecyclerQuizNavAdapter
import edu.SpaceLearning.SpaceEnglish.Listeners.InteractionActivityFragmentsListener
import edu.SpaceLearning.SpaceEnglish.Listeners.RecyclerItemQuizListener
import edu.SpaceLearning.SpaceEnglish.QuizFiles.QuizCategoriesSubFragment
import edu.SpaceLearning.SpaceEnglish.R
import edu.SpaceLearning.SpaceEnglish.UtilsClasses.Constants
import edu.SpaceLearning.SpaceEnglish.UtilsClasses.Utils
import edu.SpaceLearning.SpaceEnglish.databinding.FragmentNavQuizBinding

// Android imports

/**
 * A fragment that handles quiz navigation within the SpaceEnglish app.
 * This fragment includes UI components for selecting quiz categories,
 * configuring quiz parameters, and enabling/disabling quiz category buttons
 * based on user scores.
 */
class QuizNavFragment  : Fragment() {
    // Binding variable for FragmentNavQuizBinding
    private lateinit var binding: FragmentNavQuizBinding

    // Listener for interaction events with the parent activity
    private var interactionListener: InteractionActivityFragmentsListener? = null


    /**
     * Called to create the view hierarchy associated with the fragment.
     * @param inflater The LayoutInflater object that can be used to inflate any views in the fragment.
     * @param container The parent view that the fragment's UI should be attached to.
     * @param savedInstanceState This fragment is being re-constructed from a previous saved state.
     * @return Return the View for the fragment's UI.
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_nav_quiz, container, false)
    }

    /**
     * Called immediately after onCreateView() has returned, providing a view hierarchy for this fragment.
     * @param view The View returned by onCreateView().
     * @param savedInstanceState If non-null, this fragment is being re-constructed from a previous saved state.
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Initialize view binding
        binding = FragmentNavQuizBinding.bind(view)

        // Initialize UI components and set up RecyclerView
        setUpTvMaxQuestionPerQuiz()
        setUpRecyclerView()
    }

    /**
     * Initializes the AutoCompleteTextView for selecting maximum questions per quiz.
     * Configures the adapter and handles item selection events.
     */
    private fun setUpTvMaxQuestionPerQuiz() {
        // Set initial value from Utils.maxQuestionsPerQuiz
        binding.tvMaxQuestionsPerQuiz.setText(Utils.maxQuestionsPerQuiz.toString())
        // Update Utils.maxQuestionsPerQuiz on item selection
        binding.tvMaxQuestionsPerQuiz.onItemClickListener =
            OnItemClickListener { parent: AdapterView<*>?, view1: View?, position: Int, id: Long ->
                val selectedValue = binding.tvMaxQuestionsPerQuiz.text.toString()
                Utils.maxQuestionsPerQuiz = selectedValue.toInt()
            }
        // Set adapter for AutoCompleteTextView
        val maxQuestionsList =
            resources.getStringArray(R.array.quiz_option_max_questions_per_quizzes)
        val adapter =
            ArrayAdapter(requireActivity(), android.R.layout.simple_list_item_1, maxQuestionsList)
        binding.tvMaxQuestionsPerQuiz.setAdapter(adapter)

    }

    /**
     * Sets up RecyclerView with adapter and layout manager for displaying quiz categories.
     */
    private fun setUpRecyclerView() {
        val adapter = RecyclerQuizNavAdapter(Utils.itemRecyclerQuizNavList, object : RecyclerItemQuizListener {
            override fun onQuizItemClicked(position: Int) {
                interactionListener?.onSetRequiredCategoryFragmentQuiz(
                    QuizCategoriesSubFragment.getInstance(
                        Constants.categoryNamesList[position]
                    )
                )
            }

        })
        binding.recyclerViewQuizNav.also {
            it.adapter = adapter
            it.layoutManager = LinearLayoutManager(requireActivity())
            it.setHasFixedSize(true)
        }

    }

    /**
     * Called when the fragment is first attached to its context.
     * @param context The context to which the fragment is attached.
     */
    override fun onAttach(context: Context) {
        super.onAttach(context)
        // Ensure the context implements InteractionActivityFragmentsListener
        if (context is InteractionActivityFragmentsListener) {
            interactionListener = context
        }
    }

    /**
     * Called when the fragment is no longer attached to its context.
     */
    override fun onDetach() {
        super.onDetach()
        // Release the interaction listener to avoid memory leaks
        interactionListener = null
    }
}
